### PR TITLE
[YUNIKORN-1871] Improving test coverage for whole user/group enforcement feature

### PR DESCRIPTION
**Pull request by FrankYang0529**
### What is this PR for?

Add test cases for functions:
* checkLimitResource
* checkLimitMaxApplications
* checkLimits
* IncreaseTrackedResource
* DecreaseTrackedResource
* Headroom

### What type of PR is it?
* [ ] - Bug Fix
* [X] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?

https://issues.apache.org/jira/browse/YUNIKORN-1871

### How should this be tested?

Run `make test`.

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by FrankYang0529** [comment]
@manirajv06, this PR is ready for review. Thanks.

**Comment by manirajv06** [comment]
@FrankYang0529 Thanks for your contribution. 

I can see tests being repeated again. For example, https://github.com/apache/yunikorn-core/blob/89090713894b93e1dd35c4ed39ac0dbaa0ff63b3/pkg/common/configs/configvalidator_test.go#L652C22-L652C22 is duplicate of https://github.com/apache/yunikorn-core/commit/68a82820f0d7ff82f4d5d9b44629ae3660a184e1#diff-b51101ffb2b9d6516306e0b62bdc45428e951201eb6c4289b1c0247edccb0b6b. So please check whether test exists already. If already exists, see if you can add anymore on top of it to ensure it is more robust. Otherwise, create a new one.

I looked at the first case and observed this repeatition. So please merge duplicate methods to improve further as you have might have added a new case which is not covered currently.

**Comment by FrankYang0529** [comment]
> I can see tests being repeated again. For example, https://github.com/apache/yunikorn-core/blob/89090713894b93e1dd35c4ed39ac0dbaa0ff63b3/pkg/common/configs/configvalidator_test.go#L652C22-L652C22 is duplicate of [68a8282#diff-b51101ffb2b9d6516306e0b62bdc45428e951201eb6c4289b1c0247edccb0b6b](https://github.com/apache/yunikorn-core/commit/68a82820f0d7ff82f4d5d9b44629ae3660a184e1#diff-b51101ffb2b9d6516306e0b62bdc45428e951201eb6c4289b1c0247edccb0b6b). So please check whether test exists already. If already exists, see if you can add anymore on top of it to ensure it is more robust. Otherwise, create a new one.

Hi @manirajv06, thanks for the review. If you think they are duplicated, why we still put the "Ensure leaf queue user/group limits are within parent queue user/group limits" case in the Jira description?

I think my test cases in `configvalidator_test.go` are at a lower level - a single function. And test cases like `TestGroupLimitsWithHierarchicalQueue` in `config_test.go` are at a higher level - an integrated function including loading config and validation.

**Comment by manirajv06** [comment]
> > I can see tests being repeated again. For example, https://github.com/apache/yunikorn-core/blob/89090713894b93e1dd35c4ed39ac0dbaa0ff63b3/pkg/common/configs/configvalidator_test.go#L652C22-L652C22 is duplicate of [68a8282#diff-b51101ffb2b9d6516306e0b62bdc45428e951201eb6c4289b1c0247edccb0b6b](https://github.com/apache/yunikorn-core/commit/68a82820f0d7ff82f4d5d9b44629ae3660a184e1#diff-b51101ffb2b9d6516306e0b62bdc45428e951201eb6c4289b1c0247edccb0b6b). So please check whether test exists already. If already exists, see if you can add anymore on top of it to ensure it is more robust. Otherwise, create a new one.
> 
> Hi @manirajv06, thanks for the review. If you think they are duplicated, why we still put the "Ensure leaf queue user/group limits are within parent queue user/group limits" case in the Jira description?
> 

It is exhaustive list to ensure all cases has been captured. Scope has been defined very clearly to explain the intent of the jira. anyways, not a problem.

> I think my test cases in `configvalidator_test.go` are at a lower level - a single function. And test cases like `TestGroupLimitsWithHierarchicalQueue` in `config_test.go` are at a higher level - an integrated function including loading config and validation.

Ok, will go through in detail. Overall intention is to ensure we don't duplicate the effort.

File: pkg/scheduler/ugm/manager_test.go [hunk 1]
```diff
@@ -485,6 +505,528 @@ func TestUserGroupHeadroom(t *testing.T) {
...
+			if err != nil {
+				t.Errorf("new resource create returned error or wrong resource: error %t, res %v", err, usage)
+			}
+
+			increased := manager.IncreaseTrackedResource(queuePathParent, TestApp1, usage, tc.user)
+			assert.Equal(t, increased, true, "unable to increase tracked resource: queuepath "+queuePathParent+", app "+TestApp1+", res "+usage.String())
+
+			increased = manager.IncreaseTrackedResource(queuePathParent, TestApp2, usage, tc.user)
+			assert.Equal(t, increased, false, "should not increase tracked resource: queuepath "+queuePathParent+", app "+TestApp2+", res "+usage.String())
+		})
```
Comment (manirajv06) [inline pkg/scheduler/ugm/manager_test.go:738]: If we can assert user and group resource usage in addition to either yes or no, that gives clear picture on usage perspective as well.

**Review by manirajv06 — CHANGES_REQUESTED** [review]
Good coverage.  Added few comments. 

Also, Have we covered group selection based on limit settings configured at different queue level? For example, leaf queue might not any configured groups but parent queue might have some configuration. sometimes, none of the queue has group settings etc.

**Comment by FrankYang0529** [comment]
Hi @manirajv06, thanks for the review. I will update the PR after https://github.com/apache/yunikorn-core/pull/600 is merged, because some of my test cases like `TestMultipleGroupLimitChange` can't pass after I rebase the master branch.

File: pkg/common/configs/configvalidator_test.go [hunk 2]
```diff
@@ -646,3 +648,669 @@ func TestServiceAccountGroupName(t *testing.T) {
...
+								},
+							},
+						},
+					},
+				},
+			},
+			hasError: false,
+		},
+		{
+			name: "parent queue user group maxapplications is 0",
```
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply pkg/common/configs/configvalidator_test.go]: We check whether maxapplications is 0 in `checkLimits`, so we don't have 0 maxapplications cases in `checkLimitMaxApplications`.

Removed cases `parent queue user group maxapplications is 0`, `leaf queue user maxapplications is 0`, and `leaf queue group maxapplications is 0`.

https://github.com/apache/yunikorn-core/blob/a4b2ebb5748ac62a8b4ce3e8e6258dc7223e8eb8/pkg/common/configs/configvalidator.go#L513-L516

File: pkg/common/configs/configvalidator_test.go [hunk 3]
```diff
@@ -646,3 +648,669 @@ func TestServiceAccountGroupName(t *testing.T) {
...
+					},
+				},
+			},
+			errMsg: "",
+		},
+		{
+			name: "queue maxapplications is 0",
+			config: QueueConfig{
+				Name:            "parent",
+				MaxApplications: 0,
```
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply pkg/common/configs/configvalidator_test.go]: From [queue config document](https://yunikorn.apache.org/docs/next/user_guide/queue_config/#queues), users can't configure queue max applications as `0`, but they can choose to not set `maxapplications` on queue config.

I change the case name to `default queue maxapplications` and remove `MaxApplications: 0`, so we don't misunderstand it. Thank you.

**Comment by manirajv06** [comment]
@FrankYang0529  Now that #600  has been merged, Can you rebase and try?

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Now that #600 has been merged, Can you rebase and try?

@manirajv06, since `TestUserGroupLimitChange` and `TestMultipleGroupLimitChange` can't pass, I remove them from this PR and add them to https://github.com/apache/yunikorn-core/pull/625.

File: pkg/scheduler/ugm/manager_test.go (hunk 1 above)
Comment (FrankYang0529) [reply pkg/scheduler/ugm/manager_test.go:1062]: Add more assertions for this part. Thanks.

**Review by manirajv06 — APPROVED** [review]
I am not sure why all test cases fails. May be some cases might fail and that too depends on the expected values. If we can set the expected values based on the current code flow, it should pass. If you can re-check and include all possible cases, that would be great.

**Review by craigcondit — APPROVED** [review]
+1 LGTM.

**Comment by FrankYang0529** [comment]
> I am not sure why all test cases fails. May be some cases might fail and that too depends on the expected values. If we can set the expected values based on the current code flow, it should pass. If you can re-check and include all possible cases, that would be great.

Let's discuss in https://github.com/apache/yunikorn-core/pull/625. Thank you.

**Pull request closed without merging**